File: src/api/library.ts
/** History and favourites. */

import { call } from "./client";
import type { FavoriteEntry, HistoryEntry } from "@/types/models";

export const historyList = (limit?: number) =>
  call<HistoryEntry[]>("history_list", { limit: limit ?? null });

export const historyRemove = (id: string) =>
  call<boolean>("history_remove", { id });

export const historyClear = (onlyFailed = false) =>
  call<number>("history_clear", { onlyFailed });

export const favoritesList = () => call<FavoriteEntry[]>("favorites_list");

export const favoritesUpsert = (entry: FavoriteEntry) =>
  call<FavoriteEntry[]>("favorites_upsert", { entry });

export const favoritesRemove = (id: string) =>
  call<FavoriteEntry[]>("favorites_remove", { id });

export const favoritesContains = (url: string) =>
  call<boolean>("favorites_contains", { url });

File: src-tauri/src/commands/library.rs
//! History and favourites commands.

use tauri::State;

use crate::commands::AppState;
use crate::core::error::AppResult;
use crate::models::library::{FavoriteEntry, HistoryEntry};

#[tauri::command]
pub fn history_list(state: State<'_, AppState>, limit: Option<usize>) -> Vec<HistoryEntry> {
    state.history.list(limit)
}

#[tauri::command]
pub fn history_remove(state: State<'_, AppState>, id: String) -> AppResult<bool> {
    state.history.remove(&id)
}

/// Clear history. `only_failed` keeps successful downloads.
#[tauri::command]
pub fn history_clear(state: State<'_, AppState>, only_failed: Option<bool>) -> AppResult<usize> {
    state.history.clear(only_failed.unwrap_or(false))
}

#[tauri::command]
pub fn favorites_list(state: State<'_, AppState>) -> Vec<FavoriteEntry> {
    state.favorites.list()
}

/// Star (or re-star) a link. Keyed by URL, so this is idempotent.
#[tauri::command]
pub fn favorites_upsert(state: State<'_, AppState>, entry: FavoriteEntry) -> AppResult<Vec<FavoriteEntry>> {
    state.favorites.upsert(entry)
}

#[tauri::command]
pub fn favorites_remove(state: State<'_, AppState>, id: String) -> AppResult<Vec<FavoriteEntry>> {
    state.favorites.remove(&id)
}

#[tauri::command]
pub fn favorites_contains(state: State<'_, AppState>, url: String) -> bool {
    state.favorites.contains(&url)
}
